<!DOCTYPE html>
<html>

<head>
    <title>Create the database cross reference links</title>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes" />
    <link rel="stylesheet" type="text/css" href="https://mzkit.org/assets/js/page.css">
</head>

<body>
    <div class="container">

        <table style="width: 100%;">
            <tr>
                <td>xref {annotation}</td>
                <td style="text-align: right;">R Documentation</td>
            </tr>
        </table>

        <h2>Create the database cross reference links</h2>

        <h3>Description</h3>

        

        <h3>Usage</h3>

        <div style="width: 100%; padding-left: 3%; color: black;">
            <pre><code class='language-R' style="color: black; font-weight:normal;"><strong>xref</strong>(
    <i>chebi</i> = <span style='color: blue !important;'>NULL</span>,
    <i>KEGG</i> = <span style='color: blue !important;'>NULL</span>,
    <i>KEGGdrug</i> = <span style='color: blue !important;'>NULL</span>,
    <i>pubchem</i> = <span style='color: blue !important;'>NULL</span>,
    <i>HMDB</i> = <span style='color: blue !important;'>NULL</span>,
    <i>metlin</i> = <span style='color: blue !important;'>NULL</span>,
    <i>DrugBank</i> = <span style='color: blue !important;'>NULL</span>,
    <i>ChEMBL</i> = <span style='color: blue !important;'>NULL</span>,
    <i>chemspider</i> = <span style='color: blue !important;'>NULL</span>,
    <i>foodb</i> = <span style='color: blue !important;'>NULL</span>,
    <i>Wikipedia</i> = <span style='color: blue !important;'>NULL</span>,
    <i>lipidmaps</i> = <span style='color: blue !important;'>NULL</span>,
    <i>MeSH</i> = <span style='color: blue !important;'>NULL</span>,
    <i>ChemIDplus</i> = <span style='color: blue !important;'>NULL</span>,
    <i>MetaCyc</i> = <span style='color: blue !important;'>NULL</span>,
    <i>KNApSAcK</i> = <span style='color: blue !important;'>NULL</span>,
    <i>CAS</i> = <span style='color: blue !important;'>NULL</span>,
    <i>InChIkey</i> = <span style='color: blue !important;'>NULL</span>,
    <i>InChI</i> = <span style='color: blue !important;'>NULL</span>,
    <i>SMILES</i> = <span style='color: blue !important;'>NULL</span>,
    <i>...</i> = <span style='color: blue !important;'>NULL</span>);</code></pre>
        </div>

        <h3>Arguments</h3>

        <table>
            
<dt>chebi</dt>
<dd><p>Chemical Entities of Biological Interest (ChEBI)
 
 Chemical Entities of Biological Interest, also known as ChEBI, is 
 a chemical database and ontology of molecular entities focused on 
 'small' chemical compounds, that is part of the Open Biomedical 
 Ontologies (OBO) effort at the European Bioinformatics Institute 
 (EBI). The term "molecular entity" refers to any "constitutionally 
 or isotopically distinct atom, molecule, ion, ion pair, radical, 
 radical ion, complex, conformer, etc., identifiable as a separately 
 distinguishable entity". The molecular entities in question are 
 either products of nature or synthetic products which have potential 
 bioactivity. Molecules directly encoded by the genome, such as 
 nucleic acids, proteins and peptides derived from proteins by 
 proteolytic cleavage, are not as a rule included in ChEBI.
 
 ChEBI uses nomenclature, symbolism And terminology endorsed by the 
 International Union Of Pure And Applied Chemistry (IUPAC) And 
 nomenclature committee Of the International Union Of Biochemistry 
 And Molecular Biology (NC-IUBMB).</p></dd>


<dt>KEGG</dt>
<dd><p>KEGG (Kyoto Encyclopedia of Genes and Genomes) is 
 a collection of databases dealing with genomes, biological pathways, 
 diseases, drugs, and chemical substances. KEGG is utilized for 
 bioinformatics research and education, including data analysis in genomics, 
 metagenomics, metabolomics and other omics studies, modeling and simulation 
 in systems biology, and translational research in drug development.
 
 The KEGG database project was initiated In 1995 by Minoru Kanehisa, 
 professor at the Institute For Chemical Research, Kyoto University, 
 under the Then ongoing Japanese Human Genome Program. Foreseeing the need 
 For a computerized resource that can be used For biological interpretation 
 Of genome sequence data, he started developing the KEGG PATHWAY database. 
 It Is a collection Of manually drawn KEGG pathway maps representing 
 experimental knowledge On metabolism And various other functions Of the 
 cell And the organism. Each pathway map contains a network Of molecular 
 interactions And reactions And Is designed To link genes In the genome 
 To gene products (mostly proteins) In the pathway. This has enabled the 
 analysis called KEGG pathway mapping, whereby the gene content In the 
 genome Is compared With the KEGG PATHWAY database To examine which 
 pathways And associated functions are likely To be encoded In the 
 genome.
 
 According to the developers, KEGG Is a "computer representation" of the 
 biological system. It integrates building blocks And wiring diagrams of 
 the system—more specifically, genetic building blocks of genes And proteins, 
 chemical building blocks of small molecules And reactions, And wiring 
 diagrams of molecular interaction And reaction networks. This concept 
 Is realized in the following databases of KEGG, which are categorized 
 into systems, genomic, chemical, And health information.</p></dd>


<dt>pubchem</dt>
<dd><p>PubChem is a database of chemical molecules and 
 their activities against biological assays. The system is maintained by 
 the National Center for Biotechnology Information (NCBI), a component of 
 the National Library of Medicine, which is part of the United States National 
 Institutes of Health (NIH). PubChem can be accessed for free through a web 
 user interface. Millions of compound structures and descriptive datasets 
 can be freely downloaded via FTP. PubChem contains multiple substance descriptions 
 and small molecules with fewer than 100 atoms and 1,000 bonds. More than 
 80 database vendors contribute to the growing PubChem database.</p></dd>


<dt>HMDB</dt>
<dd><p>The Human Metabolome Database (HMDB) is a comprehensive, 
 high-quality, freely accessible, online database of small molecule metabolites 
 found in the human body. It bas been created by the Human Metabolome Project 
 funded by Genome Canada and is one of the first dedicated metabolomics 
 databases. The HMDB facilitates human metabolomics research, including the 
 identification and characterization of human metabolites using NMR spectroscopy, 
 GC-MS spectrometry and LC/MS spectrometry. To aid in this discovery process, 
 the HMDB contains three kinds of data: 1) chemical data, 2) clinical data, and 
 3) molecular biology/biochemistry data (Fig. 1–3). The chemical data includes 
 41,514 metabolite structures with detailed descriptions along with nearly 
 10,000 NMR, GC-MS and LC/MS spectra.</p></dd>


<dt>metlin</dt>
<dd><p>The METLIN Metabolite and Chemical Entity Database is the 
 largest repository of experimental tandem mass spectrometry and neutral loss data 
 acquired from standards. The tandem mass spectrometry data on over 870,000 molecular 
 standards (as of August 16, 2022) is provided to facilitate the identification 
 of chemical entities from tandem mass spectrometry experiments. In addition to the 
 identification of known molecules, it is also useful for identifying unknowns using 
 its similarity searching technology. All tandem mass spectrometry data comes from 
 the experimental analysis of standards at multiple collision energies and in both 
 positive and negative ionization modes.
 
 METLIN serves as a data management system to assist in metabolite And chemical 
 entity identification by providing public access to its repository of comprehensive 
 MS/MS And neutral loss data. METLIN's annotated list of molecular standards include 
 metabolites and other chemical entities, searching METLIN can be done based on a 
 molecule's tandem mass spectrometry data, neutral loss masses, precursor mass, 
 chemical formula, and structure within the METLIN website. Each molecule is linked
 to outside resources such as the Kyoto Encyclopedia of Genes and Genomes (KEGG) for 
 further reference and inquiry. The METLIN database was developed and is maintained 
 solely by the Siuzdak laboratory at The Scripps Research Institute.</p></dd>


<dt>DrugBank</dt>
<dd><p>The DrugBank database is a comprehensive, freely accessible, 
 online database containing information on drugs and drug targets created and maintained 
 by the University of Alberta and The Metabolomics Innovation Centre located in Alberta, 
 Canada. As both a bioinformatics and a cheminformatics resource, DrugBank combines 
 detailed drug (i.e. chemical, pharmacological and pharmaceutical) data with 
 comprehensive drug target (i.e. sequence, structure, and pathway) information. DrugBank 
 has used content from Wikipedia; Wikipedia also often links to Drugbank, posing 
 potential circular reporting issues.</p></dd>


<dt>ChEMBL</dt>
<dd><p>ChEMBL or ChEMBLdb is a manually curated chemical database of 
 bioactive molecules with drug inducing properties. It is maintained by the European 
 Bioinformatics Institute (EBI), of the European Molecular Biology Laboratory (EMBL), 
 based at the Wellcome Trust Genome Campus, Hinxton, UK.</p>

<p>The database, originally known As StARlite, was developed by a biotechnology company 
 called Inpharmatica Ltd. later acquired by Galapagos NV. The data was acquired For 
 EMBL In 2008 With an award from The Wellcome Trust, resulting In the creation Of the 
 ChEMBL chemogenomics group at EMBL-EBI, led by John Overington.</p></dd>


<dt>Wikipedia</dt>
<dd><p>Wikipedia is a free-content online encyclopedia written and 
 maintained by a community of volunteers, collectively known as Wikipedians, through
 open collaboration and using a wiki-based editing system called MediaWiki. Wikipedia 
 is the largest and most-read reference work in history. It has consistently been one 
 of the 10 most popular websites in the world, and, as of 2023, ranks as the 4th most 
 viewed website by Semrush. Founded by Jimmy Wales and Larry Sanger on January 15, 
 2001, it is hosted by the Wikimedia Foundation, an American nonprofit organization.</p></dd>


<dt>lipidmaps</dt>
<dd><p>LIPID MAPS (Lipid Metabolites and Pathways Strategy) is a web 
 portal designed to be a gateway to Lipidomics resources. The resource has spearheaded 
 a classification of biological lipids, dividing them into eight general categories. 
 LIPID MAPS provides standardised methodologies for mass spectrometry analysis of lipids, e.g.</p>

<p>LIPID MAPS has been cited As evidence Of a growing appreciation Of the study Of lipid
 metabolism And the rapid development And standardisation Of the lipidomics field</p>

<p>Key LIPID MAPS resources include:</p>

<p>1. LIPID MAPS Structure Database (LMSD) - a database Of structures And annotations Of 
    biologically relevant lipids, containing over 46000 different lipids. The paper 
    describing this resource has, according To PubMed, been cited more than 200 times.
 2. LIPID MAPS In-Silico Structure Database (LMISSD) - a database Of computationally 
    predicted lipids generated by expansion Of headgroups For commonly occurring lipid 
    classes
 3. LIPID MAPS Gene/Proteome Database (LMPD) - a database Of genes And gene products 
    which are involved In lipid metabolism
    
 Tools available from LIPID MAPS enable scientists To identify likely lipids In their 
 samples from mass spectrometry data, a common method To analyse lipids In biological 
 specimens. In particular, LipidFinder enables analysis Of MS data. Tutorials And 
 educational material On lipids are also available at the site.
 
 In January 2020, LIPID MAPS became an ELIXIR service.</p></dd>


<dt>MeSH</dt>
<dd><p>Medical Subject Headings (MeSH) is a comprehensive controlled 
 vocabulary for the purpose of indexing journal articles and books in the life sciences. 
 It serves as a thesaurus that facilitates searching. Created and updated by the 
 United States National Library of Medicine (NLM), it is used by the MEDLINE/PubMed 
 article database and by NLM's catalog of book holdings. MeSH is also used by 
 ClinicalTrials.gov registry to classify which diseases are studied by trials registered 
 in ClinicalTrials.</p>

<p>MeSH was introduced In the 1960S, With the NLM's own index catalogue and the subject
 headings of the Quarterly Cumulative Index Medicus (1940 edition) as precursors. 
 The yearly printed version of MeSH was discontinued in 2007; MeSH is now available 
 only online. It can be browsed and downloaded free of charge through PubMed. Originally 
 in English, MeSH has been translated into numerous other languages and allows 
 retrieval of documents from different origins.</p></dd>


<dt>ChemIDplus</dt>
<dd><p>ChemIDplus (Chemical Identification Plus Database)
 
 ChemIDplus was a dictionary Of over 400,000 chemicals (names, synonyms, And structures). 
 ChemIDplus includes links To NLM And other databases And resources, including links To 
 federal, state And international agencies. NLM makes a subset Of ChemIDplus data available 
 For download. The ChemIDplus Subset does Not include the Structure Or the toxicity data 
 available from the NLM web versions Of the database.</p></dd>


<dt>MetaCyc</dt>
<dd><p>The MetaCyc database is one of the largest metabolic pathways 
 and enzymes databases currently available. The data in the database is manually 
 curated from the scientific literature, and covers all domains of life. MetaCyc has 
 extensive information about chemical compounds, reactions, metabolic pathways and 
 enzymes. The data have been curated from more than 58,000 publications.</p>

<p>MetaCyc has been designed For multiple types Of uses. It Is often used As an extensive 
 online encyclopedia Of metabolism. In addition, MetaCyc Is used As a reference data 
 Set For computationally predicting the metabolic network Of organisms from their sequenced 
 genomes; it has been used To perform pathway predictions For thousands Of organisms, 
 including those In the BioCyc Database Collection. MetaCyc Is also used In metabolic 
 engineering And metabolomics research.</p>

<p>MetaCyc includes mini reviews For pathways And enzymes that provide background information 
 As well As relevant literature references. It also provides extensive data On individual 
 enzymes, describing their subunit Structure, cofactors, activators And inhibitors, 
 substrate specificity, And, When available, kinetic constants. MetaCyc data On metabolites 
 includes chemical structures, predicted Standard energy Of formation, And links To external 
 databases. Reactions In MetaCyc are presented In a visual display that includes the 
 structures Of all components. The reactions are balanced And include EC numbers, reaction 
 direction, predicted atom mappings that describe the correspondence between atoms In 
 the reactant compounds And the product compounds, And computed Gibbs free energy.</p>

<p>All objects In MetaCyc are clickable And provide easy access To related objects. For example,
 the page For L-lysine lists all Of the reactions In which L-lysine participates, 
 As well As the enzymes that catalyze them And pathways In which these reactions take place.</p></dd>


<dt>KNApSAcK</dt>
<dd><p>KNApSAcK: A Comprehensive Species-Metabolite Relationship Database</p></dd>


<dt>CAS</dt>
<dd><p>A CAS Registry Number (also referred to as CAS RN or informally CAS 
 Number) is a unique identification number assigned by the Chemical Abstracts Service 
 (CAS) in the US to every chemical substance described in the open scientific literature. 
 It includes all substances described since 1957, plus some substances from as far 
 back as the early 1800s. It is a chemical database that includes organic and inorganic 
 compounds, minerals, isotopes, alloys, mixtures, and nonstructurable materials 
 (UVCBs, substances of unknown or variable composition, complex reaction products, 
 or biological origin). CAS RNs are generally serial numbers (with a check digit), 
 so they do not contain any information about the structures themselves the way 
 SMILES and InChI strings do.
 
 The registry maintained by CAS Is an authoritative collection Of disclosed chemical 
 substance information. It identifies more than 204 million unique organic And inorganic 
 substances And 70 million protein And DNA sequences, plus additional information about 
 Each substance. It Is updated With around 15,000 additional New substances daily.
 A collection Of almost 500 thousand CAS registry numbers are made available under a 
 CC BY-NC license at ACS Commons Chemistry.</p></dd>


<dt>InChIkey</dt>
<dd><p>The International Chemical Identifier (InChI /ˈɪntʃiː/ IN-chee 
 or /ˈɪŋkiː/ ING-kee) is a textual identifier for chemical substances, designed to
 provide a standard way to encode molecular information and to facilitate the search 
 for such information in databases and on the web. Initially developed by the International
 Union of Pure and Applied Chemistry (IUPAC) and National Institute of Standards and 
 Technology (NIST) from 2000 to 2005, the format and algorithms are non-proprietary. 
 Since May 2009, it has been developed by the InChI Trust, a nonprofit charity from 
 the United Kingdom which works to implement and promote the use of InChI.</p></dd>


<dt>InChI</dt>
<dd><p>The International Chemical Identifier (InChI /ˈɪntʃiː/ IN-chee 
 or /ˈɪŋkiː/ ING-kee) is a textual identifier for chemical substances, designed to
 provide a standard way to encode molecular information and to facilitate the search 
 for such information in databases and on the web. Initially developed by the International
 Union of Pure and Applied Chemistry (IUPAC) and National Institute of Standards and 
 Technology (NIST) from 2000 to 2005, the format and algorithms are non-proprietary. 
 Since May 2009, it has been developed by the InChI Trust, a nonprofit charity from 
 the United Kingdom which works to implement and promote the use of InChI.</p></dd>


<dt>SMILES</dt>
<dd><p>The simplified molecular-input line-entry system (SMILES) is 
 a specification in the form of a line notation for describing the structure of 
 chemical species using short ASCII strings. SMILES strings can be imported by 
 most molecule editors for conversion back into two-dimensional drawings or three-dimensional 
 models of the molecules.
 
 The original SMILES specification was initiated In the 1980S. It has since been 
 modified And extended. In 2007, an open standard called OpenSMILES was developed 
 In the open source chemistry community.</p></dd>


<dt>chemspider</dt>
<dd><p>ChemSpider is a freely accessible online database of 
 chemicals owned by the Royal Society of Chemistry. It contains information on more than 100 
 million molecules from over 270 data sources, each of them receiving a unique identifier 
 called ChemSpider Identifier.</p></dd>


<dt>foodb</dt>
<dd><p>FooDB (The Food Database) is a freely available, open-access 
 database containing chemical (micronutrient and macronutrient) composition data on 
 common, unprocessed foods. It also contains extensive data on flavour and aroma 
 constituents, food additives as well as positive and negative health effects 
 associated with food constituents. The database contains information on more than
 28,000 chemicals found in more than 1000 raw or unprocessed food products. The 
 data in FooDB was collected from many sources including textbooks, scientific journals, 
 on-line food composition or nutrient databases, flavour and aroma databases 
 and various on-line metabolomic databases. This literature-derived information has
 been combined with experimentally derived data measured on thousands of compounds 
 from more than 40 very common food products through the Alberta Food Metabolome 
 Project which is led by David S. Wishart. Users are able to browse through the FooDB 
 data by food source, name, descriptors or function. Chemical structures and molecular 
 weights for compounds in FooDB may be searched via a specialized chemical structure 
 search utility. Users are able to view the content of FooDB using two different 
 “Viewing” options: FoodView, which lists foods by their chemical compounds, or ChemView,
 which lists chemicals by their food sources. Knowledge about the precise chemical 
 composition of foods can be used to guide public health policies, assist food companies 
 with improved food labelling, help dieticians prepare better dietary plans, 
 support nutraceutical companies with their submissions of health claims and guide 
 consumer choices with regard to food purchases.</p></dd>


<dt>KEGGdrug</dt>
<dd><p>KEGG DRUG is a comprehensive drug information resource for 
 approved drugs in Japan, USA and Europe, unified based on the chemical structure 
 and/or the chemical component of active ingredients. Each KEGG DRUG entry is 
 identified by the D number and associated with KEGG original annotations including 
 therapeutic targets, drug metabolism, and other molecular interaction network 
 information.</p></dd>

        </table>


        <h3 style="display: none;">Details</h3>

        

        <h3 style="display: block">Authors</h3>

        BioNovoGene

        <h3>Value</h3>

        this function returns data object of type <a href="/vignettes/clr/BioNovoGene/BioDeep/Chemistry/MetaLib/CrossReference/xref.html">xref</a>.<h4>clr value class</h4><ul><li><a href="/vignettes/clr/BioNovoGene/BioDeep/Chemistry/MetaLib/CrossReference/xref.html">xref</a></li></ul>

        <h3 style="display: block;">Examples</h3>

        <pre><code id="example_r"> # an example of create metabolite annotation data
 # for 'ATP'.
 
 let xrefs = annotation::xref(
     KEGG = 'C00002',
     CAS = '56-65-5',
     pubchem = '3304',
     chebi = '15422',
     KNApSAcK = 'C00001491'
 );</code></pre>

        <hr />
        <div style="text-align: center;">[Package <em>annotation</em> version 1.0.0.0 <a
                href="../annotation.html">Index</a>]
        </div>
    </div>
</body>

<script type="text/javascript" src="https://mzkit.org/assets/js/R_syntax.js"></script>
<script type="text/javascript" src="https://mzkit.org/assets/js/highlights.js"></script>
<script type="text/javascript">r_highlights("example_r");</script>

</html>